service/history/timerQueueActiveProcessor.go

```diff
@@ -604,15 +647,70 @@ Update_History_Loop:
 			}
 		}
 
-		if e := msBuilder.AddTimeoutWorkflowEvent(); e == nil {
-			// If we failed to add the event that means the workflow is already completed.
-			// we drop this timeout event.
-			return nil
+		retryBackoffInterval := msBuilder.GetRetryBackoffDuration(getTimeoutErrorReason(workflow.TimeoutTypeStartToClose))
+		if retryBackoffInterval == common.NoRetryBackoff {
+			if e := msBuilder.AddTimeoutWorkflowEvent(); e == nil {
+				// If we failed to add the event that means the workflow is already completed.
+				// we drop this timeout event.
+				return nil
+			}
+
+			// We apply the update to execution using optimistic concurrency.  If it fails due to a conflict than reload
+			// the history and try the operation again.
+			err = t.updateWorkflowExecution(context, msBuilder, false, true, nil, nil)
+			if err != nil {
+				if err == ErrConflict {
+					continue Update_History_Loop
+				}
+			}
+			return err
+		}
+
+		// workflow timeout, but a retry is needed, so we do continue as new to retry
+		startEvent, err := getWorkflowStartedEvent(t.historyService.historyMgr, domainID, workflowExecution.GetWorkflowId(), workflowExecution.GetRunId())
+		if err != nil {
+			return err
+		}
+
+		startAttributes := startEvent.WorkflowExecutionStartedEventAttributes
+		continueAsnewAttributes := &workflow.ContinueAsNewWorkflowExecutionDecisionAttributes{
+			WorkflowType: startAttributes.WorkflowType,
+			TaskList:     startAttributes.TaskList,
+			RetryPolicy:  startAttributes.RetryPolicy,
+			Input:        startAttributes.Input,
+			ExecutionStartToCloseTimeoutSeconds: startAttributes.ExecutionStartToCloseTimeoutSeconds,
+			TaskStartToCloseTimeoutSeconds:      startAttributes.TaskStartToCloseTimeoutSeconds,
+			BackoffStartIntervalInSeconds:       common.Int32Ptr(int32(retryBackoffInterval.Seconds())),
+		}
+		domainEntry, err := getActiveDomainEntryFromShard(t.shard, &domainID)
+		if err != nil {
+			return err
+		}
+		_, continueAsNewBuilder, err := msBuilder.AddContinueAsNewEvent(common.EmptyEventID, domainEntry, startAttributes.GetParentWorkflowDomain(), continueAsnewAttributes)
+		if err != nil {
+			return err
+		}
+
+		tBuilder := t.historyService.getTimerBuilder(&context.workflowExecution)
+		var transferTasks, timerTasks []persistence.Task
+		tranT, timerT, err := getDeleteWorkflowTasksFromShard(t.shard, domainID, tBuilder)
```

Comment:
Can we test to make sure we are not creating 2 separate delete tasks on workflow timeout.

Reply:
workflowTimeout does not go through this path.